active-rfcs/0000-vuex-module-api.md

```diff
@@ -0,0 +1,289 @@
+
+  computed: counter.mapGetters(['...']),
+
+  methods: counter.mapActions(['...'])
+})
+```
+
+In another module:
+
+```js
+import { Actions } from 'vuex'
+import { counter } from '@/store/modules/counter'
+
+// class-style syntax
+class AnotherActions extends Actions {
+  // create module context of counter module
+  counter = counter.context(this.store)
+
+  incrementCounter() {
+    // commit `increment` mutation in counter
+    this.counter.commit('increment')
+  }
+}
+
+// object-style syntax
+const anotherActions = {
+  incrementCounter() {
+    const counterCtx = counter.context(this)
+    counterCtx.commit('increment')
+  }
+}
+```
+
+# Motivation
+
+<!--
+Why are we doing this? What use cases does it support? What is the expected
+outcome?
+
+Please focus on explaining the motivation so that if this RFC is not accepted,
+the motivation could be used to develop alternative solutions. In other words,
+enumerate the constraints you are trying to solve without coupling them too
+closely to the solution you have in mind.
+-->
+
+To ensure type safety when referring module assets from components or any other place in apps.
+Currently, it is not typed at all because:
+
+1. `mapXXX` helpers cannot know the type of module assets,
+2. It is impossible to deal with namespace string on type level and
+3. typing store instance is too complicated as it has nested structure and needs string concatnation on type level.
+
+To solve these problem, we utilize a module object which is fully typed by class-style syntax. And automatically resolve
+namespace string in the module object.
+
+# Detailed design
+
+<!--
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Vue to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+-->
+
+## Module object
+
+Providing `Module` constructor to create a module object. It receives module options which is the same shape with the current plain object module:
+
+```js
+import { Module } from 'vuex'
+
+// create counter module
+export const counter = new Module({
+  namespaced: true,
+
+  // module assets
+  state: ...,
+  getters: ...,
+  mutations: ...,
+  actions: ...,
+
+  // nested modules
+  modules: {
+    ...
+  }
+})
+```
+
+The module object can be passed to the store constructore's first argument. If the first argument is the module object, the second argument will be store options:
+
+```js
+import { Store } from 'vuex'
+import { counter } from './modules/counter'
+
+export default new Store(counter, {
+  strict: true
+})
+```
+
+After registering a module object, it will store the registered module path and namespace string under the hood so that the user do not have to specify it by their hand.
+The module object has to be unique in a store because of this behaviour.
```

Comment:
Do we need to check uniqueness in runtime library? It seems hard for new comers to remember that store could be only used once.

Reply:
Yeah, we should provide warning in that case.